linkml/generators/pydanticgen/array.py

```diff
@@ -0,0 +1,396 @@
+                imports=_AnyShapeArrayImports,
+            )
+
+    def labeled_shape(self, array: ArrayExpression) -> SlotResult:
+        """
+        Constrained shapes using :func:`pydantic.conlist`
+
+        TODO:
+        - preservation of aliases
+        - (what other metadata is allowable on labeled dimensions?)
+        """
+        # generate dimensions from inside out and then format
+        range = self.dtype
+        for dimension in reversed(array.dimensions):
+            range = self._labeled_dimension(dimension, range).annotation
+
+        return SlotResult(annotation=range, imports=_ConListImports)
+
+    def mixed_shape(self, array: ArrayExpression) -> SlotResult:
+        """
+        Mixture of labeled dimensions with a max or min (or both) shape for anonymous dimensions
+        """
+        if array.exact_number_dimensions is not None:
+            if array.exact_number_dimensions > len(array.dimensions):
+                res = SlotResult(
+                    annotation=self._list_of_lists(array.exact_number_dimensions - len(array.dimensions), self.dtype)
+                )
+            elif array.exact_number_dimensions == len(array.dimensions):
+                # equivalent to labeled shape
+                return self.labeled_shape(array)
+            else:
+                raise ValueError(
+                    "if exact_number_dimensions is provided, it must be greater than the parameterized dimensions"
+                )
+
+        elif array.maximum_number_dimensions is not None and not array.maximum_number_dimensions:
+            # unlimited n dimensions, so innermost is AnyShape with dtype
+            res = self.any_shape(with_inner_union=True)
+
+            if array.minimum_number_dimensions and array.minimum_number_dimensions > len(array.dimensions):
+                # some minimum anonymous dimensions but unlimited max dimensions
+                res += self._list_of_lists(array.minimum_number_dimensions - len(array.dimensions), res.annotation)
+
+        elif array.minimum_number_dimensions and array.maximum_number_dimensions is None:
+            raise ValueError(
+                (
+                    "Cannot specify a minimum_number_dimensions while maximum is None while using labeled dimensions - "
+                    "either use exact_number_dimensions > len(dimensions) for extra anonymous dimensions or set "
+                    "maximum_number_dimensions explicitly to False for unbounded dimensions"
+                )
+            )
+        elif array.maximum_number_dimensions:
+            initial_min = array.minimum_number_dimensions if array.minimum_number_dimensions is not None else 0
+            dmin = max(len(array.dimensions), initial_min) - len(array.dimensions)
+            dmax = array.maximum_number_dimensions - len(array.dimensions)
+
+            res = self.anonymous_shape(ArrayExpression(minimum_number_dimensions=dmin, maximum_number_dimensions=dmax))
```

Comment:
I believe it is possible here for `dmax = 0`, i.e., `array.maximum_number_dimensions` is specified and equals the length of the labeled dimensions. This results in an `ArrayExpression` where `maximum_number_dimensions==0` which is equal to False and I think it is interpreted as such (unlimited number of dimensions) in `self.anonymous_shape`. Maybe this case is caught above though.

Reply:
```suggestion
            # dmin can be 0. if dmax = 1, then res.annotation = Union[dtype, List[dtype]]
            res = self.anonymous_shape(ArrayExpression(minimum_number_dimensions=dmin, maximum_number_dimensions=dmax))
```

Note that it is also possible here for `dmin = 0` but tracing through the code, I think that is OK and works out well.

Reply:
will add to test cases and evaluate monday